src/middleware.js

```diff
@@ -15,6 +14,160 @@ import { handleCarBlock } from './handlers/car-block.js'
+    const tokenMetadata = await accounting.getTokenMetadata(authToken)
+    if (tokenMetadata) {
+      await env.AUTH_TOKEN_METADATA.put(authToken, serializeTokenMetadata(tokenMetadata))
+      return tokenMetadata
+    } else {
+      return null
+    }
+  }
+}
+
+/**
+ * @type {import('./bindings.js').RateLimits}
+ */
+const RateLimits = {
+  create: ({ env }) => ({
+    check: async (cid, options) => {
+      const authToken = await getAuthorizationTokenFromRequest(options)
+      if (authToken) {
+        console.log(`found token ${authToken}, looking for content commitment`)
+        const tokenMetadata = await getTokenMetadata(env, authToken)
+
+        if (tokenMetadata) {
+          if (tokenMetadata.invalid) {
+            // this means we know about the token and we know it's invalid, so we should just use the CID rate limit
+            return checkRateLimitForCID(env, cid)
+          } else {
+            // TODO at some point we should enforce user configurable rate limits and origin matching
+            // but for now we just serve all valid token requests
+            return RATE_LIMIT_EXCEEDED.NO
+          }
+        } else {
+          // we didn't get any metadata - for now just use the top level rate limit
+          // this means token based requests will be subject to normal rate limits until the data propagates
+          return checkRateLimitForCID(env, cid)
+        }
+      } else {
+        // no token, use normal rate limit
+        return checkRateLimitForCID(env, cid)
+      }
+    }
+  })
+}
+
+/**
+ * @type {import('./bindings.js').Accounting}
+ */
+const Accounting = {
+  create: ({ serviceURL }) => ({
+    record: async (cid, options) => {
+      console.log(`using ${serviceURL} to record a GET for ${cid} with options`, options)
+    },
+
+    getTokenMetadata: async () => {
+      // TODO I think this needs to check the content claims service (?) for any claims relevant to this token
+      // TODO do we have a plan for this? need to ask Hannah if the indexing service covers this?
+      return null
+    }
+  })
+}
+
+/**
+ *
+ * @param {Pick<Request, 'headers'>} request
+ * @returns string
+ */
+async function getAuthorizationTokenFromRequest (request) {
+  // TODO this is probably wrong
+  const authToken = request.headers.get('Authorization')
+  return authToken
+}
+
+/**
+ *
+ * @type {import('@web3-storage/gateway-lib').Middleware<IpfsUrlContext, IpfsUrlContext, Environment>}
+ */
+export function withRateLimits (handler) {
+  return async (request, env, ctx) => {
+    if (!env.FF_RATE_LIMITER_ENABLED) {
+      console.warn('rate limiting disabled')
+      return handler(request, env, ctx)
+    }
+
+    const { dataCid } = ctx
+
+    const rateLimits = RateLimits.create({ env })
+    const isRateLimitExceeded = await rateLimits.check(dataCid, request)
+
+    if (isRateLimitExceeded === RATE_LIMIT_EXCEEDED.YES) {
+      // TODO should we record this?
```

Comment:
**suggestion:** I think we don't need to—we do get some basic HTTP logging for free on Cloudflare, right?